source/tests/system/nidaqmx_driver_api_tests.cpp

```diff
@@ -1439,10 +1478,178 @@ TEST_F(NiDAQmxDriverApiTests, ChannelWithDoneEventRegisteredTwice_RunCompleteFin
   read_analog_f64(FINITE_SAMPLE_COUNT, FINITE_SAMPLE_COUNT);
   RegisterDoneEventResponse response1, response2;
   read_stream(reader_context1, *reader1, response1);
-  read_stream(reader_context2, *reader2, response2);
 
   EXPECT_EQ(DAQMX_SUCCESS, response1.status());
-  EXPECT_EQ(DONE_EVENT_ALREADY_REGISTERED_ERROR, response2.status());
+  EXPECT_THROW({
+    try {
+      read_stream(reader_context2, *reader2, response2);
+    }
+    catch (const client::grpc_driver_error& ex) {
+      expect_driver_error(ex, DONE_EVENT_ALREADY_REGISTERED_ERROR);
+      throw;
+    }
+  }, client::grpc_driver_error);
+}
+
+TEST_F(NiDAQmxDriverApiTests, EveryNSamplesEventRegisteredWithWrongEventType_ReadStream_NotSupportedByDeviceErrorReceived)
+{
+  const auto N_SAMPLES = 10UL;
+  create_ai_voltage_chan(0.0, 1.0);
+  ::grpc::ClientContext reader_context;
+  auto reader = register_every_n_samples_event(reader_context, N_SAMPLES, EveryNSamplesEventType::EVERY_N_SAMPLES_EVENT_TYPE_TRANSFERRED_FROM_BUFFER);
+
+  RegisterEveryNSamplesEventResponse response;
+  EXPECT_THROW({
+    try {
+      read_stream(reader_context, *reader, response);
+    }
+    catch (const client::grpc_driver_error& ex) {
+      expect_driver_error(ex, EVERY_N_SAMPS_TRANSFERRED_FROM_BUFFER_EVENT_NOT_SUPPORTED_BY_DEVICE_ERROR);
+      throw;
+    }
+  }, client::grpc_driver_error);
+}
+
+TEST_F(NiDAQmxDriverApiTests, EveryNSamplesEventRegisteredWithWrongEventType_Finish_NotSupportedByDeviceErrorReceived)
+{
+  const auto N_SAMPLES = 10UL;
+  create_ai_voltage_chan(0.0, 1.0);
+  ::grpc::ClientContext reader_context;
+  auto reader = register_every_n_samples_event(reader_context, N_SAMPLES, EveryNSamplesEventType::EVERY_N_SAMPLES_EVENT_TYPE_TRANSFERRED_FROM_BUFFER);
+
+  EXPECT_THROW({
+    try {
+      auto status = reader->Finish();
+      client::raise_if_error(status, reader_context);
+    }
+    catch (const client::grpc_driver_error& ex) {
+      expect_driver_error(ex, EVERY_N_SAMPS_TRANSFERRED_FROM_BUFFER_EVENT_NOT_SUPPORTED_BY_DEVICE_ERROR);
+      throw;
+    }
+  }, client::grpc_driver_error);
+}
+
+TEST_F(NiDAQmxDriverApiTests, EveryNSamplesEventRegisteredWithWrongEventType_Cancel_NoErrors)
+{
+  const auto N_SAMPLES = 10UL;
+  create_ai_voltage_chan(0.0, 1.0);
+  ::grpc::ClientContext reader_context;
+  auto reader = register_every_n_samples_event(reader_context, N_SAMPLES, EveryNSamplesEventType::EVERY_N_SAMPLES_EVENT_TYPE_TRANSFERRED_FROM_BUFFER);
+
+  reader_context.TryCancel();
+}
+
+TEST_F(NiDAQmxDriverApiTests, AsyncEveryNSamplesEventRegistered_OnEventReadAnalogF64_EventAndReadResponsesAreReceived)
+{
+  const auto N_SAMPLES = 10UL;
+  const auto N_READS = 10UL;
+  const auto EVENT_START_CALL_TAG = (void*)1;
+  const auto EVENT_READ_INITIAL_METADATA_TAG = (void*)2;
+  const auto EVENT_READ_TAG = (void*)3;
+  const auto READ_ANALOG_F64_TAG = (void*)4;
+  create_ai_voltage_chan(0.0, 1.0);
+  ::grpc::ClientContext event_context;
+  ::grpc::CompletionQueue completion_queue;
```

Comment:
While writing the register/unregister tests, I found that if the test function returns or throws while there are pending async operations, it may corrupt the stack. These tests should shut down the completion queue on exit to ensure that it has no dangling references to stack variables.

Reply:
Fixed by using RAII to shut down the completion queue:
```
auto shut_down_on_exit = make_scope_exit([&]{ shut_down_completion_queue(completion_queue); });
```